File: articles/cryptographic-algorithms-and-secure-communication/calculators/python/secure_communication_calculator.py
#!/usr/bin/env python3
"""Small secure-communication governance calculator.

This calculator scores governance readiness. It does not implement production
cryptography.
"""

from __future__ import annotations

from pathlib import Path
import csv

ARTICLE_ROOT = Path(__file__).resolve().parents[2]
OUTPUT = ARTICLE_ROOT / "calculators" / "outputs" / "secure_communication_calculator.csv"


def score_case(values: dict[str, float]) -> float:
    weights = {
        "threat_model": 0.12,
        "protocols": 0.10,
        "keys": 0.14,
        "validation": 0.10,
        "integrity": 0.10,
        "authentication": 0.10,
        "randomness": 0.08,
        "secret_storage": 0.10,
        "rotation": 0.08,
        "implementation_review": 0.08,
    }
    return round(100 * sum(weights[key] * values[key] for key in weights), 3)


def main() -> None:
    OUTPUT.parent.mkdir(parents=True, exist_ok=True)
    rows = [
        {
            "case": "standard secure channel",
            "score": score_case({
                "threat_model": 0.86,
                "protocols": 0.88,
                "keys": 0.82,
                "validation": 0.90,
                "integrity": 0.86,
                "authentication": 0.84,
                "randomness": 0.84,
                "secret_storage": 0.82,
                "rotation": 0.78,
                "implementation_review": 0.80,
            }),
        },
        {
            "case": "legacy manual transfer",
            "score": score_case({
                "threat_model": 0.36,
                "protocols": 0.28,
                "keys": 0.24,
                "validation": 0.18,
                "integrity": 0.34,
                "authentication": 0.28,
                "randomness": 0.30,
                "secret_storage": 0.22,
                "rotation": 0.16,
                "implementation_review": 0.20,
            }),
        },
    ]

    with OUTPUT.open("w", newline="", encoding="utf-8") as handle:
        writer = csv.DictWriter(handle, fieldnames=["case", "score"])
        writer.writeheader()
        writer.writerows(rows)

    for row in rows:
        print(f"{row['case']}: {row['score']}")
    print(OUTPUT)


if __name__ == "__main__":
    main()

File: articles/cryptographic-algorithms-and-secure-communication/calculators/r/secure_communication_calculator.R
# Base R secure-communication governance calculator.

args <- commandArgs(trailingOnly = FALSE)
file_arg <- grep("^--file=", args, value = TRUE)

if (length(file_arg) > 0) {
  script_path <- normalizePath(sub("^--file=", "", file_arg[1]), mustWork = TRUE)
  article_root <- normalizePath(file.path(dirname(script_path), "..", ".."), mustWork = TRUE)
} else {
  article_root <- getwd()
}

output_dir <- file.path(article_root, "calculators", "outputs")
if (!dir.exists(output_dir)) {
  dir.create(output_dir, recursive = TRUE)
}

score_case <- function(threat_model, protocols, keys, validation, integrity, authentication, randomness, secret_storage, rotation, implementation_review) {
  100 * (
    0.12 * threat_model +
      0.10 * protocols +
      0.14 * keys +
      0.10 * validation +
      0.10 * integrity +
      0.10 * authentication +
      0.08 * randomness +
      0.10 * secret_storage +
      0.08 * rotation +
      0.08 * implementation_review
  )
}

rows <- data.frame(
  case = c("standard secure channel", "legacy manual transfer"),
  score = c(
    score_case(0.86, 0.88, 0.82, 0.90, 0.86, 0.84, 0.84, 0.82, 0.78, 0.80),
    score_case(0.36, 0.28, 0.24, 0.18, 0.34, 0.28, 0.30, 0.22, 0.16, 0.20)
  )
)

write.csv(rows, file.path(output_dir, "r_secure_communication_calculator.csv"), row.names = FALSE)
print(rows)
